conda/models/match_spec.py

```diff
@@ -1021,13 +1020,16 @@ def match(self, other):
         else:
             # assert ChannelMatch('pkgs/free').match('defaults') is False
             # assert ChannelMatch('defaults').match('pkgs/free') is True
+
+            # This should have been the following
+            # self._raw_value.name == _other_val.canonical_name or self._raw_value == _other_val
+            # but users may rely on this exact behaviour.
+            # For eg: conda install -c file::/path/to/distr distr::pkg as mentioned in
+            # https://anaconda.org/distr
             return self._raw_value.name in (_other_val.name, _other_val.canonical_name)
 
     def __str__(self):
-        try:
-            return "%s" % self._raw_value.name
-        except AttributeError:
-            return "%s" % self._raw_value
+        return str(self._raw_value)
```

Comment:
This might be breaking a few tests:

```
FAILED tests/models/test_match_spec.py::test_hash_merge_with_name - AssertionError: assert 'https://conda.anaconda.org/conda-forge::python=1.2.3[md5=deadbeef]' == 'conda-forge::python=1.2.3[md5=deadbeef]'
  - conda-forge::python=1.2.3[md5=deadbeef]
  + https://conda.anaconda.org/conda-forge::python=1.2.3[md5=deadbeef]
```

I think we should return the name when it comes from the `channel_alias`, and the full URL otherwise.